runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -1025,13 +1025,26 @@ protected void fireTimer(TimerData timerData) {
     checkArgument(namespace instanceof WindowNamespace);
     BoundedWindow window = ((WindowNamespace) namespace).getWindow();
     timerInternals.onFiredOrDeletedTimer(timerData);
+    Instant effectiveOutputTimestamp;
+
+    if (timerData.getDomain() == TimeDomain.EVENT_TIME) {
+      effectiveOutputTimestamp = timerData.getOutputTimestamp();
+    } else {
+      // Flink does not set a watermark hold for the timer's output timestamp, and previous to
+      // https://github.com/apache/beam/pull/17262 processing time timers did not correctly emit
+      // elements at their output timestamp.  In this case we need to continue doing the wrong thing
+      // and using the output watermark rather than the firing timestamp.  Once flink correctly sets
+      // a  watermark hold for the output timestamp, this should be changed back.
+      effectiveOutputTimestamp = timerInternals.currentOutputWatermarkTime();
```

Comment:
Should this be `currentInputWatermarkTime()`?

Reply:
should it?  Didn't we say we wanted to use output watermark instead?  I can certainly change it to be the input watermark, but I figured I'd make it slightly less wrong while I was in here.

Reply:
Yes, but as I understood, the reason for this change is to preserve the _current_ behavior, which sets input watermark as output timestamp, or am I misunderstanding it?

Reply:
I'd say more to preserve correctness (no late data) than the current behavior.  I don't use Flink so I don't have much skin the game here, if we want to keep it as it was I can certainly keep the existing behavior completely.

Reply:
I would say we should (temporarily) ignore the output timestamp. Hence
```java
  timer.withOutputTimestamp(...).set(..)
 ```
 and
 ```java
   timer.set(...)
 ```
 should work exactly the same.

Reply:
that's basically what it's doing now, the question is what timestamp do you want to use for the output timestamp.  Using the input watermark allows you to end up with a timestamp outside of the current window.  The output watermark won't have that problem since the cleanup timer will hold the watermark.

Reply:
Sure, but without specifying the output timestamp, we currently use the input watermark, right? Yes, we ideally should do a follow-up to fix that globally to output watermark, but currently it works that way and there IMHO should not be a difference in output timestamp when using `.withOutputTimestamp()` and without, if we do not support it correctly.

Reply:
I'm not sure what you're saying at this point?  The output timestamp (either set explicitly or implicitly) is completely ignored currently and with this change on Flink.

Reply:
Yes, sorry. :+1: